extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/RequestScopeHandlerInterceptor.java

```diff
@@ -0,0 +1,41 @@
+package io.quarkus.grpc.runtime.supports;
+
+import io.grpc.ForwardingServerCall;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ManagedContext;
+
+public class RequestScopeHandlerInterceptor implements ServerInterceptor {
+
+    private final ManagedContext reqContext;
+
+    public RequestScopeHandlerInterceptor() {
+        reqContext = Arc.container().requestContext();
+    }
+
+    @Override
+    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+            Metadata headers,
+            ServerCallHandler<ReqT, RespT> next) {
+
+        boolean managed = !reqContext.isActive();
+        if (managed) {
+            reqContext.activate();
+        }
+
+        return next.startCall(new ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT>(call) {
+            @Override
+            public void close(Status status, Metadata trailers) {
+                super.close(status, trailers);
+                if (managed) {
+                    reqContext.deactivate();
```

Comment:
@cescoffier Do you expect this to be called from a different thread than the one that called `reqContext.activate()`?

Reply:
It could be the case if the method is annotated with @Blocking or if the implementation calls "onComplete" or "onError" from another thread.

Reply:
Well, I don't have enough knowledge about gRPC and its interceptors but `deactivate()` should be always called on the same thread (to cleanup the thread locals) and we should also call `destroy()` to destroy the beans correctly when the request processing is over. In reactive routes we capture the state and add an end handler: https://github.com/quarkusio/quarkus/blob/master/extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/RouteHandler.java#L46-L52

Reply:
Hum... that is a bit tricky, I need another approach as we cannot guarantee this.
Even if I capture the context, that context may be used by another invocation, which may even complete before the first call (and so destroy the beans created for the first call). 

There is no "routing context" in gRPC which makes things a bit more tricky. 

I'm wondering how you handle the context reuse in reactive routes?

Reply:
> I'm wondering how you handle the context reuse in reactive routes?

What exactly do you mean? In any case, the question should be "how do we handle the context reuse..." ;-)

Reply:
If I understand correctly, the request scope is thread dependant, which means that it stored in a thread-local attached to the thread "activating" it. As a consequence, deactivation should be done from the same thread, so clean up everything.

Everything work when the thread is used to handle a single concurrent call. 
But it may not be the case in the reactive world.

For example in the reactive routes, your method is called on the IO thread (event loop). That thread may be used to handle multiple concurrent requests. If one of these requests enables the request scope, what does it means for the other requests handled (concurrently) on the same event loop?  If two concurrent requests both enable the request scope, and one terminates before the other one, how can you be sure that it does not close the request scope of the other (still progressing) request? 

I just found a way to have something similar to the routing context in gRPC, but I need to be sure that we would not step on other's toes.

Reply:
> If I understand correctly, the request scope is thread dependant, which means that it stored in a thread-local attached to the thread "activating" it.

Yes.

> But it may not be the case in the reactive world.

Yes, that's why MP Context Propagation was created (although it does not cover all possible use cases).

> For example in the reactive routes, your method is called on the IO thread (event loop). That thread may be used to handle multiple concurrent requests.

The request context is deactivated (thread local removed) right after a route method completes so it should not be a problem (a IO thread may not execute multiple route methods at the same time). If you do some async inside the route method MP CP makes a snapshot of the context before the async action is performed and this snapshot is propagated further.

Reply:
Ok, so it requires MP Context propagation in these cases, or you got things messed up. 

So, I should be able to use the same machinery and restore the right (execution) context. I just need to store it and propagate it.

Reply:
> Ok, so it requires MP Context propagation in these cases, or you got things messed up.

Yes, if you need the context to be propagated. In fact, you can leverage the ArC APIs directly but that's not something I would recommend ;-).

> So, I should be able to use the same machinery...

I think so.

Reply:
I'm happy to say that I've found a simple solution -> Just capturing the context.